### ISSUE TITLE
Extensions Design

### ISSUE THREAD

**Issue by RobBagby** [Azure/acs-engine#599]
# Microsoft Azure Container Service Engine - Extensions

Extensions in acs-engine will provide an easy way for acs-engine users to add pre-packaged functionality into their cluster.  For example, an extension could configure a monitoring solution on an ACS cluster.  The user would not need to know the details of how to install the monitoring solution.  Rather, the user would simply add the extension into the extensionsProfile section of the template.

At a high level, extensions simply take advantage of a couple of ARM constructs: linked templates and usually custom script extensions.  The typical extension will simply add a Microsoft.Resources/deployments resource to the ARM template that points to a linked template.  The linked template will typically have a custom script extension that points at a script file.

This issue outlines the design for extensions. 

# extensionsProfile

The extensionsProfile contains the extensions that the cluster will install. The following illustrates a template with a hello-world-k8s extension.

``` json
{ 
  ...
  "extensionsProfile": [
    {
        "name": "hello-world-k8s",
        "version": "v1",
        "extensionParameters": "parameters",
        "rootURL": "http://mytestlocation.com/hello-world-k8s/"
    }
  ]
}
```

|Name|Required|Description|
|---|---|---|
|name|yes|the name of the extension.  This has to exactly match the name of a folder under the extensions folder|
|version|yes|the version of the extension.  This has to exactly match the name of the folder under the extension name folder|
|extensionParameters|optional|extension parameters may be required by extensions.  The format of the parameters is also extension dependant.|
|rootURL|optional|url to the root location of extensions.  The rootURL must have an extensions child folder that follows the extensions convention.  The rootURL is mainly used for testing purposes.|

# rootURL
You normally would not provide a rootURL.  The extensions are normally loaded from the extensions folder in GitHub.  However, you may specify the rootURL when testing a new extension.  The rootURL must adhere to the extensions conventions.  For example, in order to use an Azure Storage account to test an extension named extension-one, you would do the following:
- Create a storage account.  For the purposes of this example, we will call it 'mystorageaccount'
- Create a blob container called 'extensions'
- Under 'extensions', create a folder called 'extension-one'
- Under 'extension-one', create a folder called 'v1'
- Under 'v1', upload your files (see Required Extension Files) 
- Set the rootURL to: 'https://mystorageaccount.blob.core.windows.net/'

# masterProfile, agentPoolProfiles
As noted at the beginning of this issue, extensions typically run a script on one or more nodes.  The extensions array in the masterProfile and agentPoolProfiles define which pools have the script run on them.

``` json
{
  "masterProfile": {
      "count": 3,
      "dnsPrefix": "dnsprefix",
      "vmSize": "Standard_D2_v2",
      "osType": "Linux",
      "firstConsecutiveStaticIP": "10.240.255.5",
      "extensions": [
        { 
          "name": "hello-world-k8s", 
          "singleOrAll": "single",
        }
     ]
  },
  "agentPoolProfiles": [
      {
        "name": "agent128",
        "count": 3,
        "vmSize": "Standard_D2_v2",
        "availabilityProfile": "AvailabilitySet",
        "storageProfile": "StorageAccount",
        "diskSizesGB": [128, 128, 128, 128],
        "extensions": [
           { 
             "name": "hello-world-k8s", 
             "singleOrAll": "all",
           }
        ]
      }
  ],
  "extensionsProfile": [
    {
        "name": "hello-world-k8s",
        "version": "v1",
        "extensionParameters": "parameters"
    }
  ]
}
```

|Name|Required|Description|
|---|---|---|
|name|yes|The name of the extension.  This must match the name in the extensionsProfile|
|singleOrAll|yes|Defines whether the extension is installed on a single node or every node in the pool.|
|template|optional|This will allow for the ability to install different scripts on different pools|

# Extension Files

In order to install an extension, there are four typical files - constraints.json, template.json, template-link.json and EXTENSION-NAME.sh. Following is a description of each file.

|File Name|Description|
|-----------------------------|---|
|constraints.json |Defines constraints such as which orchestrators are supported by the extension (Swarm, Dcos, or Kubernetes) or any dependencies such as requiring a specific storageProfile|
|template.json               |The ARM template used to deploy the extension|
|template-link.json          |The ARM template snippet which will be injected into azuredeploy.json to call template.json|
|EXTENSION-NAME.sh           |The script file that will execute on the VM itself via Custom Script Extension to perform installation of the extension|

**Comment by amanohar** [comment]
cc: @anhowe

**Comment by rjtsdl** [comment]
Just my 2 cent,

Wondering if this can leverage the existing package manager for certain orchestrators, such as Helm for K8s, or Docker Stack for Swarm. Then we can enjoy the existing great ecosystem. 

For example, in our extension, if user specify a package manager (like, Helm), and a package name (known by Helm). We can deploy on the user behalf.

**Comment by seanknox** [comment]
Agreed @rjtsdl. I bias toward leaning on Helm to handle this type of logic (for Kubernetes, at least).

**Comment by stuartleeks** [comment]
Would it work if we created a Helm extension that had parameters for the package to install? (And enable multiple instances of the same extension)

My thinking is that doing that would enable simple consumption of Helm packages on k8s, but also enable other deployments if not in Helm.

**Comment by RobBagby** [comment]
During the week of 5/15 - 5/19, several engineers from TED (Stuart Leeks, myself, Paul Bouwer, Julien Stroheker and Harold Wong) worked on extensions.  The following is an update on the work that was done and the findings of the team.

- There are challenges with regards to using linked templates to run custom script extensions on agent pools because some of the agent pools are implemented as vm scale sets.  VM scale sets do not allow you to deploy scripts in custom script extensions in linked templates.
- The team determined that the MVP for extensions was the mechanism to run a script on a single master
- Running scripts on one or all agents in a pool will be implemented in a later effort
- The team was able to implement the MVP 
- The team was able to build several extensions
  - A helm extension that installs Helm
  - A kubernetes OMS extension
  - A Sysdig extension
  - A VAMP extension on DC/OS
- We will be building a helm chart extension that will allow you to install any helm chart.  This extension will require the implementation of a dependency model - where one extension can depend upon another.  This should not be challenging, in that the use of linked templates will allow for this.

To keep in line with the API suggested by Anabhuti, we implemented an extensions array in the masterProfile:

```javascript
{
  "masterProfile": {
      "count": 3,
      "dnsPrefix": "dnsprefix",
      "vmSize": "Standard_D2_v2",
      "osType": "Linux",
      "firstConsecutiveStaticIP": "10.240.255.5",
      "extensions": [
        { 
          "name": "hello-world-k8s", 
        }
     ]
  },
   "extensionsProfile": [
    {
        "name": "hello-world-k8s",
        "version": "v1",
        "extensionParameters": "parameters"
    }
  ]
}
```

The code does a check to validate that for an extension, it must exist in the masterProfile extensions array.

We will be issuing a new PR with the updated code on 5/26 or 5/30.

**Comment by amanohar** [comment]
@stuartleeks +1 to what you mentioned about having the option is deployments are not Helm as well. This will enable the feature for DC/OS and Swarm as well in addition to more options for K8s.

@RobBagby thanks for the update. Looking forward to the PR.

Also, regarding: _There are challenges with regards to using linked templates to run custom script extensions on agent pools because some of the agent pools are implemented as vm scale sets. VM scale sets do not allow you to deploy scripts in custom script extensions in linked templates._ 
Can a check be added to disable this feature for VMSS agents?

**Comment by RobBagby** [comment]
@amanohar Yes - we could add in a check.  Question for you: will all agent pools will be implemented as VMSS soon?

**Comment by amanohar** [comment]
@RobBagby, K8s will continue using VM instead of VMSS agents for the foreseeable future.

**Comment by RobBagby** [comment]
@amanohar, Good to know - we can issue another PR with k8s support for agent pools then

**Comment by stuartleeks** [comment]
Can we get some comments on this? There is a PR (#687) that is sat going stale because it hasn't been reviewed.

**Comment by JackQuincy** [comment]
So I've been working on this and I realized the design was flawed. We can't use a separate custom script extension since it makes it so that we have more than one custom script on a given node which isn't supported.  @RobBagby @stuartleeks How did you get the k8s OMS extension working? Did you test it? Every time I run it I get a failure due to there being more than one extension on the VM.  But above you say you succeeded in doing it. From the conversations I've had it seems like the design is currently flawed due to this limitation. If this won't work, I'm looking at possibly moving the extensions into the existing custom scripts for nodes that have them. This will have limitations in reusablility etc. But should meet the needs our team has about installing software on the VMs outside of the orchestrator.

**Comment by stuartleeks** [comment]
IIRC: we re-use the existing custom script extension in a new template and set a dependency on the initial script extension completing. When we re-use the script extension we point it at a different script.

**Comment by stuartleeks** [comment]
I told @RobBagby I'd rebase the PR, but I ended up having some time off for family matters. Happy to take a look at that tomorrow...

**Comment by JackQuincy** [comment]
Ah that is why it had the same name. Makes sense I'll try and see if that unblocks. I had tried to name them more intuitively
